tests/extra_api/socket_isolation.c

```diff
@@ -0,0 +1,277 @@
+    assert(strncmp(buf, HELLO_MSG, strlen(HELLO_MSG)) == 0);
+
+    /*
+     * Socket close
+     */
+
+    sleep(1);
+
+    rc = close(sock_in);
+    assert(rc == 0);
+    rc = close(sock_in2);
+    assert(rc == 0);
+    rc = close(sock);
+    assert(rc == 0);
+    rc = close(sock2);
+    assert(rc == 0);
+    rc = close(sock3);
+    assert(rc == 0);
+}
+
+static void client(const char *server_ip)
+{
+    char buf[64];
+    struct sockaddr_in addr;
+    ssize_t len;
+    int sock;
+    int sock2;
+    int val = SO_XLIO_ISOLATE_SAFE;
+    int valdef = SO_XLIO_ISOLATE_DEFAULT;
+    int rc;
+
+    sock = socket(AF_INET, SOCK_STREAM, 0);
+    assert(sock >= 0);
+    sock2 = socket(AF_INET, SOCK_STREAM, 0);
+    assert(sock2 >= 0);
+
+    rc = setsockopt(sock, SOL_SOCKET, SO_XLIO_ISOLATE, &val, sizeof(val));
+    assert(rc == 0);
+    rc = setsockopt(sock, SOL_SOCKET, SO_XLIO_ISOLATE, &valdef, sizeof(valdef));
+    assert(rc == -1);
+    assert(errno == EINVAL);
+
+    memset(&addr, 0, sizeof(addr));
+    addr.sin_family = AF_INET;
+    addr.sin_addr.s_addr = inet_addr(server_ip);
+    addr.sin_port = htons(8080);
+
+    rc = connect(sock, (struct sockaddr *)&addr, sizeof(addr));
+    assert(rc == 0);
+
+    addr.sin_port = htons(8081);
+    rc = connect(sock2, (struct sockaddr *)&addr, sizeof(addr));
+    assert(rc == 0);
+
+    rc = setsockopt(sock2, SOL_SOCKET, SO_XLIO_ISOLATE, &val, sizeof(val));
+    assert(rc == -1);
+    assert(errno == EINVAL);
+
+    int xlio_ring_fds[3];
+    int xlio_ring_fds2[3];
+    rc = xlio_api->get_socket_rings_fds(sock, xlio_ring_fds, ARRAY_SIZE(xlio_ring_fds));
+    assert(rc == 1);
+    rc = xlio_api->get_socket_rings_fds(sock2, xlio_ring_fds2, ARRAY_SIZE(xlio_ring_fds2));
+    assert(rc == 1);
+    assert(xlio_ring_fds[0] != xlio_ring_fds2[0]);
+
+    len = write(sock, HELLO_MSG, sizeof(HELLO_MSG));
+    assert(len > 0);
+
+    do {
+        len = read(sock, buf, sizeof(buf));
+    } while (len == -1 && errno == EINTR);
+    assert(len > 0);
+    assert(len == sizeof(HELLO_MSG));
+    assert(strncmp(buf, HELLO_MSG, strlen(HELLO_MSG)) == 0);
+
+    sleep(1);
+
+    rc = close(sock);
+    assert(rc == 0);
+    rc = close(sock2);
+    assert(rc == 0);
+}
+
+static void usage(const char *name)
+{
+    printf("Usage: %s <-s|-c> <server-ip>\n", name);
+    printf(" -s         server mode\n");
+    printf(" -c         client mode\n");
+    printf(" server-ip  IPv4 address to listen/connect to\n");
+    exit(1);
+}
+
+int main(int argc, char **argv)
+{
+    if (argc != 3) {
+        usage(argc > 0 ? argv[0] : "a.out");
+    }
+
+    xlio_api = xlio_get_api();
+    if (xlio_api == NULL) {
+        printf("Extra API not found. Run under XLIO.\n");
+        return 1;
+    }
+
+    if (strcmp(argv[1], "-s") == 0) {
+        server(argv[2]);
+    } else if (strcmp(argv[1], "-c") == 0) {
+        client(argv[2]);
+    } else {
+        usage(argv[0]);
+    }
+
+    printf("Success.\n");
+    return 0;
+}
```

Comment:
Why isn't this a part of our GTEST tests?

Reply:
manual test was easier to debug. This needs to be converted to a gtest for CI

Reply:
Will you convert it as part of this PR?

Reply:
lets review without gtest for now. I will convert it either in this PR or a separated depending on when the main logic is ready for merge